### PR TITLE
WooCommerce: Dashboard / Setup / NUX: Persist taxes-opt-out, shipping-opt-out and i'm-finished to the server

### PR DESCRIPTION
**Pull request by allendav**
This PR adds actions to persist some of the user's dashboard/setup/nux choices (opt out of tax setup, opt out of shipping setup and i'm finished) to the server for #13612

It also does a little bit of rearranging of views to support a setup-address-view (coming in a subsequent PR) and a setup-progress-view (also coming in a subsequent PR) for #14775 

To test:
- npm run test-client client/extensions/woocommerce/state/sites/setup-choices and make sure everything passes
- Use https://developer.wordpress.com/docs/api/console/ to POST to /sites/{siteID}/calypso-preferences/woocommerce and set finished_initial_setup, opted_out_of_shipping_setup and opted_out_of_taxes_setup to 0
- Go to http://calypso.localhost:3000/store/{site} and make sure you see the task list with the shipping and taxes tasks
- Click on the "I won't be shipping" link. Make sure the shipping task goes away
- Force a refresh. Make sure the shipping task stays away
- Click on the "I'm not charging sales tax" link. Make sure the tax task goes away
- Force a refresh. Make sure the tax task stays away
- Click on the "I'm finished" button at the bottom of the page
- Make sure the dashboard content is replaced with the "no-orders-yet" view aka the view that starts with "Congratulations! Now that your setup is complete you can spread the word and get those first orders rolling in."

### CODE REVIEW

**Comment by allendav** [comment]
@jeffstieler - you're an optional reviewer

File: client/extensions/woocommerce/app/dashboard/setup-tasks.js [hunk 1]
```diff
@@ -1,74 +1,91 @@
...
 
-	// TODO - replace with props mapped from state when this info becomes available in state
-	static defaultProps = {
-		storeHasBeenCustomized: true,
-		storeHasProducts: false,
-		storePaymentsAreSetUp: false,
-		storeShippingIsSetUp: false,
-		storeTaxesAreSetUp: false,
+	state = {
+		showShippingTask: this.props.loading || ! this.props.optedOutOfShippingSetup,
```
Comment (allendav) [inline client/extensions/woocommerce/app/dashboard/setup-tasks.js]: I'll take a look - thx!

**Comment by allendav** [comment]
@justinshreve wrote:

> The links like "I'm not charging any sales tax" need a cursor: pointer CSS rule applied to them.

Fixed in https://github.com/Automattic/wp-calypso/pull/15223/commits/0e2607b75640dde06fbe095f8a390d6fa8449fab

**Comment by allendav** [comment]
@justinshreve wrote:

> We are also not importing for v1, so that button and description could also be removed.

Fixed in https://github.com/Automattic/wp-calypso/pull/15223/commits/e95cee3b776e67613abd79fb56053b5b15c1d3bf and https://github.com/Automattic/wp-calypso/pull/15223/commits/589ff83dde533f4059f48f22148b55a0ac68d437

File: client/extensions/woocommerce/app/dashboard/setup-tasks-view.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+					label={ translate( 'I\'m finished setting up' ) }
+					primary={ allTasksCompleted }
+				/>
+			</div>
+		);
+	}
+}
+
+function mapStateToProps( state ) {
+	// TODO - add test for hasProducts, paymentsAreSetUp, shippingIsSetUp and taxesAreSetUp
```
Comment (allendav) [inline client/extensions/woocommerce/app/dashboard/setup-tasks-view.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/15223/commits/839bf8f28cfca716ded1417ec5216dbaac1df544

File: client/extensions/woocommerce/app/dashboard/setup-tasks.js (hunk 1 above)
Comment (allendav) [inline client/extensions/woocommerce/app/dashboard/setup-tasks.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/15223/commits/5b766e2f1cd5fc11d3bc6f0dbd547fc198cde9bb

**Comment by allendav** [comment]
@justinshreve - all comments are in - ty! please let me know if further changes are needed

File: client/extensions/woocommerce/app/dashboard/setup-tasks.js [hunk 3]
```diff
@@ -1,92 +1,114 @@
...
 	};
 
-	// TODO - replace with props mapped from state when this info becomes available in state
-	static defaultProps = {
-		storeHasBeenCustomized: true,
-		storeHasProducts: false,
-		storePaymentsAreSetUp: false,
-		storeShippingIsSetUp: false,
-		storeTaxesAreSetUp: false,
+	constructor( props ) {
```
Comment (belcherj) [inline client/extensions/woocommerce/app/dashboard/setup-tasks.js]: UI state is preferable, but should not hold this up

File: client/extensions/woocommerce/state/action-types.js [hunk 4]
```diff
@@ -38,6 +38,8 @@ export const WOOCOMMERCE_SETTINGS_PRODUCTS_REQUEST = 'WOOCOMMERCE_SETTINGS_PRODU
 export const WOOCOMMERCE_SETTINGS_PRODUCTS_REQUEST_SUCCESS = 'WOOCOMMERCE_SETTINGS_PRODUCTS_REQUEST_SUCCESS';
 export const WOOCOMMERCE_SETUP_CHOICES_REQUEST = 'WOOCOMMERCE_SETUP_CHOICES_REQUEST';
 export const WOOCOMMERCE_SETUP_CHOICES_REQUEST_SUCCESS = 'WOOCOMMERCE_SETUP_CHOICES_REQUEST_SUCCESS';
+export const WOOCOMMERCE_SETUP_CHOICE_UPDATE_REQUEST = 'WOOCOMMERCE_SETUP_CHOICE_UPDATE_REQUEST';
```
Comment (belcherj) [inline client/extensions/woocommerce/state/action-types.js]: These should be above: WOOCOMMERCE_SETUP_CHOICES_REQUEST

File: client/extensions/woocommerce/state/sites/setup-choices/actions.js [hunk 5]
```diff
@@ -9,6 +9,8 @@ import { setError } from '../status/wc-api/actions';
 import {
 	WOOCOMMERCE_SETUP_CHOICES_REQUEST,
```
Comment (belcherj) [inline client/extensions/woocommerce/state/sites/setup-choices/actions.js]: WOOCOMMERCE_SETUP_CHOICE_UPDATE_REQUEST should be above: WOOCOMMERCE_SETUP_CHOICES_REQUEST

File: client/extensions/woocommerce/state/sites/setup-choices/actions.js [hunk 6]
```diff
@@ -40,3 +42,53 @@ export const fetchSetupChoices = ( siteId ) => ( dispatch, getState ) => {
 			dispatch( setError( siteId, getAction, err ) );
 		} );
 };
+
+const updateSetupChoice = ( dispatch, siteId, key, value ) => {
+	const postAction = {
+		type: WOOCOMMERCE_SETUP_CHOICE_UPDATE_REQUEST,
+		siteId,
+		key,
+		value
```
Comment (belcherj) [inline client/extensions/woocommerce/state/sites/setup-choices/actions.js]: trailing comma

File: client/extensions/woocommerce/state/sites/setup-choices/reducer.js [hunk 7]
```diff
@@ -6,6 +6,7 @@ import { LOADING } from 'woocommerce/state/constants';
 import {
 	WOOCOMMERCE_SETUP_CHOICES_REQUEST,
 	WOOCOMMERCE_SETUP_CHOICES_REQUEST_SUCCESS,
+	WOOCOMMERCE_SETUP_CHOICE_UPDATE_REQUEST_SUCCESS,
```
Comment (belcherj) [inline client/extensions/woocommerce/state/sites/setup-choices/reducer.js]: alpha

File: client/extensions/woocommerce/state/sites/setup-choices/test/actions.js [hunk 8]
```diff
@@ -76,4 +90,328 @@ describe( 'actions', () => {
...
+
+		it( 'should dispatch an action', () => {
+			const getState = () => ( {} );
+			const dispatch = spy();
+			setTriedCustomizerDuringInitialSetup( siteId, true )( dispatch, getState );
+			expect( dispatch ).to.have.been.calledWith( {
+				type: WOOCOMMERCE_SETUP_CHOICE_UPDATE_REQUEST,
+				siteId,
+				key: 'tried_customizer_during_initial_setup',
+				value: true
```
Comment (belcherj) [inline client/extensions/woocommerce/state/sites/setup-choices/test/actions.js]: trailing

File: client/extensions/woocommerce/state/sites/setup-choices/test/actions.js [hunk 9]
```diff
@@ -76,4 +90,328 @@ describe( 'actions', () => {
...
+
+		it( 'should dispatch an action', () => {
+			const getState = () => ( {} );
+			const dispatch = spy();
+			setSetStoreAddressDuringInitialSetup( siteId, true )( dispatch, getState );
+			expect( dispatch ).to.have.been.calledWith( {
+				type: WOOCOMMERCE_SETUP_CHOICE_UPDATE_REQUEST,
+				siteId,
+				key: 'set_store_address_during_initial_setup',
+				value: true
```
Comment (belcherj) [inline client/extensions/woocommerce/state/sites/setup-choices/test/actions.js]: trailing

Comment (allendav) [inline client/extensions/woocommerce/state/sites/setup-choices/test/actions.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/15223/commits/004863a3c5f48ce79a2d0ea93ba459112536c873

File: client/extensions/woocommerce/state/sites/setup-choices/test/actions.js (hunk 8 above)
Comment (allendav) [inline client/extensions/woocommerce/state/sites/setup-choices/test/actions.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/15223/commits/004863a3c5f48ce79a2d0ea93ba459112536c873

File: client/extensions/woocommerce/state/sites/setup-choices/reducer.js (hunk 7 above)
Comment (allendav) [inline client/extensions/woocommerce/state/sites/setup-choices/reducer.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/15223/commits/004863a3c5f48ce79a2d0ea93ba459112536c873

File: client/extensions/woocommerce/state/sites/setup-choices/actions.js (hunk 6 above)
Comment (allendav) [inline client/extensions/woocommerce/state/sites/setup-choices/actions.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/15223/commits/004863a3c5f48ce79a2d0ea93ba459112536c873

File: client/extensions/woocommerce/state/sites/setup-choices/actions.js (hunk 5 above)
Comment (allendav) [inline client/extensions/woocommerce/state/sites/setup-choices/actions.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/15223/commits/004863a3c5f48ce79a2d0ea93ba459112536c873

File: client/extensions/woocommerce/state/action-types.js (hunk 4 above)
Comment (allendav) [inline client/extensions/woocommerce/state/action-types.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/15223/commits/004863a3c5f48ce79a2d0ea93ba459112536c873

File: client/extensions/woocommerce/app/dashboard/setup-tasks.js [hunk 10]
```diff
@@ -1,92 +1,114 @@
...
 import React, { Component, PropTypes } from 'react';
 
 /**
  * Internal dependencies
  */
+import {
+	areSetupChoicesLoading,
+	getOptedOutOfShippingSetup,
+	getOptedOutofTaxesSetup,
+	getTriedCustomizerDuringInitialSetup
```
Comment (allendav) [inline client/extensions/woocommerce/app/dashboard/setup-tasks.js]: Fixed in https://github.com/Automattic/wp-calypso/pull/15223/commits/004863a3c5f48ce79a2d0ea93ba459112536c873

**Pull request merged**